### PR TITLE
fix(loop): serialize /new consolidation and preserve session on archival failure

### LINKED ISSUE #880: Race in /new memory consolidation can trigger concurrent writers
## Bug Description
A race exists in `AgentLoop._process_message` around the `/new` command path.

After upstream PR #823 added `_consolidating` deduplication for normal background consolidation, `/new` still starts a separate consolidation task without sharing the same serialization mechanism. If a normal consolidation is already in-flight, `/new` can run a second writer concurrently against `MEMORY.md` / `HISTORY.md`.

## Why this matters
- two consolidation writers can overlap for the same session
- `/new` clears session state before archival success is known
- task references are not retained, which can make task lifecycle tracking fragile

## Expected behavior
- consolidation for a given session should be serialized across both normal path and `/new`
- `/new` should archive first, then clear session only on successful archival
- background consolidation tasks should be strongly referenced while in-flight

## Regression origin
This became relevant after the consolidation guard introduced by upstream PR #823 because `/new` remained outside that guard/serialization path.

## Repro (high level)
1. Trigger normal consolidation by exceeding `memory_window`
2. Immediately send `/new`
3. Observe overlapping consolidation executions for one session

I have a PR ready with tests and a minimal fix.

### PR DESCRIPTION
**Pull request by Athemis**
## Summary
- serialize memory consolidation per session across both normal background path and `/new`
- keep strong references to in-flight consolidation tasks until completion
- require successful archival result before `/new` clears the live session (prevents data loss when consolidation returns failure)
- switch consolidation to `self.context.memory` and add regression tests for `/new` + concurrency/failure behavior

## Context
This addresses a race related to the guard introduced in #823: normal consolidation was deduplicated, but `/new` could still start a concurrent writer for the same session.

Follow-up fix included from Athemis/nanobot-redux@80c8c28b50c80a13642538f8909c48ecf5576fd0 to ensure archival failure is propagated to `/new` behavior.

Closes #880.

### CODE REVIEW

File: nanobot/agent/loop.py [hunk 1]
```diff
@@ -293,56 +313,75 @@ async def _process_message(
...
+                    await self._consolidate_memory(temp_session, archive_all=True)
+            except Exception as e:
+                logger.error("/new archival failed for {}: {}", session.key, e)
+                return OutboundMessage(
+                    channel=msg.channel,
+                    chat_id=msg.chat_id,
+                    content="Could not start a new session because memory archival failed. Please try again.",
+                )
+
             session.clear()
```
[the remark this answers is not in the document]
Comment (Athemis) [reply nanobot/agent/loop.py:357]: Addressed P1 review feedback:
- `/new` now requires successful archival before session clear
- consolidation returns success/failure and `/new` preserves session on failure
- regression coverage updated in `tests/test_consolidate_offset.py`

**Comment by Athemis** [comment]
Thanks a lot for the feedback.

> **One edge case to consider:**
> 
> The `_consolidation_locks` dict grows unboundedly — one lock per unique session key, never cleaned up. For long-running gateway instances with many unique sessions this could be a slow leak. A simple mitigation would be to remove the lock when consolidation completes and the lock is not held:
> 
> ```python
> if not lock.locked():
>     self._consolidation_locks.pop(session.key, None)
> ```

This is fixed in c1b5e8c8d29a2225a1f82a36a93da5a0b61d702f. However in a slightly more complex way:

Using only `if not lock.locked():` is not sufficient, because an `asyncio.Lock` can be briefly “unlocked” between owners while there are still queued waiters for that same lock. If we remove the lock from _consolidation_locks during that window, a new caller can create a new lock for the same session key. Then some coroutines continue waiting on the old lock, while others proceed with the new lock - breaking per-session serialization and potentially allowing concurrent consolidation again.
That is why we also check for waiters and only prune when:
- `not lock.locked()`
- and there are no pending waiters.
This avoids unbounded lock-map growth without reintroducing the race condition.
